### DISCUSSION TITLE
Deploy---> alex---> dev---> Lern---> LernDataProducts

### DISCUSSION THREAD

**Discussion by srikanth-6520** [Sunbird-Ed/Community#525] [Bugs]
Hi Team, 
we are getting an error, while deploying the LerndataProducts , below we have provided jenkins console.

Started by user [Kolluri Srikanth](https://jenkins.alex-sb.dev.verlab.io/user/srikanthk)
Rebuilds build #[45](https://jenkins.alex-sb.dev.verlab.io/job/Deploy/job/alex/job/dev/job/Lern/job/LernDataProducts/45)
Checking out git https://github.com/srikanth-6520/data-products/ into /var/lib/jenkins/workspace/Deploy/alex/dev/Lern/LernDataProducts@script/cbccd55647f4734a7e1118d32120ff1b9550c27f489edc4740d07be2247508b0 to read pipelines/deploy/lern-dataproducts/Jenkinsfile
The recommended git tool is: NONE
No credentials specified
 > git rev-parse --resolve-git-dir /var/lib/jenkins/workspace/Deploy/alex/dev/Lern/LernDataProducts@script/cbccd55647f4734a7e1118d32120ff1b9550c27f489edc4740d07be2247508b0/.git # timeout=10
Fetching changes from the remote Git repository
 > git config remote.origin.url https://github.com/srikanth-6520/data-products/ # timeout=10
Using shallow fetch with depth 1
Fetching upstream changes from https://github.com/srikanth-6520/data-products/
 > git --version # timeout=10
 > git --version # 'git version 2.17.1'
 > git fetch --tags --progress --depth=1 -- https://github.com/srikanth-6520/data-products/ +refs/heads/*:refs/remotes/origin/* # timeout=10
 > git rev-parse origin/release-5.1.0^{commit} # timeout=10
Checking out Revision 630ca3978348aadcedf0fc009119dd346a496a91 (origin/release-5.1.0)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f 630ca3978348aadcedf0fc009119dd346a496a91 # timeout=10
Commit message: "Update main.yml"
 > git rev-list --no-walk 66c9c501a2f9da81290fc0e6d963a5202563ed75 # timeout=10
Loading library deploy-conf@heads/shared-lib
Attempting to resolve heads/shared-lib from remote references...
 > git --version # timeout=10
 > git --version # 'git version 2.17.1'
 > git ls-remote -- https://github.com/project-sunbird/sunbird-devops.git # timeout=10
The recommended git tool is: NONE
No credentials specified
 > git rev-parse --resolve-git-dir /var/lib/jenkins/workspace/Deploy/alex/dev/Lern/LernDataProducts@libs/b4ac4d1235c67cfedda9d7bf40936b35bc6929460ef19107abe1b183748d4145/.git # timeout=10
Fetching changes from the remote Git repository
 > git config remote.origin.url https://github.com/project-sunbird/sunbird-devops.git # timeout=10
Using shallow fetch with depth 1
Fetching upstream changes from https://github.com/project-sunbird/sunbird-devops.git
 > git --version # timeout=10
 > git --version # 'git version 2.17.1'
 > git fetch --no-tags --progress --depth=1 -- https://github.com/project-sunbird/sunbird-devops.git +refs/heads/*:refs/remotes/origin/* refs/heads/shared-lib # timeout=10
Checking out Revision 778da7ce7f97efd003eed98071921a837d209343 (778da7ce7f97efd003eed98071921a837d209343)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f 778da7ce7f97efd003eed98071921a837d209343 # timeout=10
Commit message: "fix: move variables to be accessible across both try blocks"
 > git rev-list --no-walk 778da7ce7f97efd003eed98071921a837d209343 # timeout=10
[Pipeline] Start of Pipeline
[Pipeline] node
Running on [Jenkins](https://jenkins.alex-sb.dev.verlab.io/computer/(built-in)/) in /var/lib/jenkins/workspace/Deploy/alex/dev/Lern/LernDataProducts
[Pipeline] {
[Pipeline] stage
[Pipeline] { (checkout public repo)
[Pipeline] echo
Found .git folder. Clearing it..
[Pipeline] sh
+ git clean -fxd
Removing ansible/lern-data-products-1.0-distribution.tar.gz
Removing lern_data_products_artifacts.zip:release-5.0.1
Removing metadata.json
Skipping repository private/
Removing private@tmp/
[Pipeline] checkout
The recommended git tool is: NONE
No credentials specified
 > git rev-parse --resolve-git-dir /var/lib/jenkins/workspace/Deploy/alex/dev/Lern/LernDataProducts/.git # timeout=10
Fetching changes from the remote Git repository
 > git config remote.origin.url https://github.com/srikanth-6520/data-products/ # timeout=10
Using shallow fetch with depth 1
Fetching upstream changes from https://github.com/srikanth-6520/data-products/
 > git --version # timeout=10
 > git --version # 'git version 2.17.1'
 > git fetch --tags --progress --depth=1 -- https://github.com/srikanth-6520/data-products/ +refs/heads/*:refs/remotes/origin/* # timeout=10
 > git rev-parse origin/release-5.1.0^{commit} # timeout=10
Checking out Revision 630ca3978348aadcedf0fc009119dd346a496a91 (origin/release-5.1.0)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f 630ca3978348aadcedf0fc009119dd346a496a91 # timeout=10
Commit message: "Update main.yml"
[Pipeline] }
[Pipeline] // stage
[Pipeline] ansiColor
[Pipeline] {

[Pipeline] ansiColor
[Pipeline] {

[Pipeline] sh
+ echo Deploy/alex/dev/Lern/LernDataProducts
[Pipeline] sh
+ echo Deploy/alex/dev/Lern/LernDataProducts
[Pipeline] sh
+ echo Deploy/alex/dev/Lern/LernDataProducts
[Pipeline] stage
[Pipeline] { (parameter checks)
[Pipeline] copyArtifacts
Copied 2 artifacts from "[ArtifactUpload » dev » Lern » LernDataProducts](https://jenkins.alex-sb.dev.verlab.io/job/ArtifactUpload/job/dev/job/Lern/job/LernDataProducts/)" build number [3](https://jenkins.alex-sb.dev.verlab.io/job/ArtifactUpload/job/dev/job/Lern/job/LernDataProducts/3/)
[Pipeline] sh
+ jq -r .artifact_name metadata.json
[Pipeline] sh
+ jq -r .node_name metadata.json
[Pipeline] echo
artifact_version not specified, using the artifact_version specified in metadata.json. 
[Pipeline] sh
+ jq -r .artifact_version metadata.json
[Pipeline] }
[Pipeline] // stage
[Pipeline] }

[Pipeline] // ansiColor
[Pipeline] stage
[Pipeline] { (get artifact)
[Pipeline] sh
+ pwd
[Pipeline] ansiColor
[Pipeline] {

[Pipeline] echo
Option chosen is JenkinsJob, using the artifacts copied
[Pipeline] }

[Pipeline] // ansiColor
[Pipeline] }
[Pipeline] // stage
[Pipeline] stage
[Pipeline] { (deploy artifact)
[Pipeline] sh
+ unzip lern_data_products_artifacts.zip:release-5.0.1
Archive:  lern_data_products_artifacts.zip:release-5.0.1
  inflating: lern-data-products-1.0-distribution.tar.gz  
+ mv lern-data-products-1.0-distribution.tar.gz ansible
[Pipeline] echo
{env=dev, module=Lern, jobName=LernDataProducts, absolute_job_path=ArtifactUpload/dev/Lern/LernDataProducts, agent=built-in, artifact_name=lern_data_products_artifacts.zip, artifact_version=release-5.0.1, currentWs=/var/lib/jenkins/workspace/Deploy/alex/dev/Lern/LernDataProducts, artifact=lern_data_products_artifacts.zip:release-5.0.1, ansiblePlaybook=--extra-vars "remote=spark" /var/lib/jenkins/workspace/Deploy/alex/dev/Lern/LernDataProducts/ansible/lpa_lern-data-products_deploy.yml, ansibleExtraArgs= --tags lern-dataproducts --vault-password-file /var/lib/jenkins/secrets/vault-pass}
[Pipeline] ansiColor
[Pipeline] {

[Pipeline] stage
[Pipeline] { (ansible-run)
[Pipeline] echo
{env=dev, module=Lern, jobName=LernDataProducts, absolute_job_path=ArtifactUpload/dev/Lern/LernDataProducts, agent=built-in, artifact_name=lern_data_products_artifacts.zip, artifact_version=release-5.0.1, currentWs=/var/lib/jenkins/workspace/Deploy/alex/dev/Lern/LernDataProducts, artifact=lern_data_products_artifacts.zip:release-5.0.1, ansiblePlaybook=--extra-vars "remote=spark" /var/lib/jenkins/workspace/Deploy/alex/dev/Lern/LernDataProducts/ansible/lpa_lern-data-products_deploy.yml, ansibleExtraArgs= --tags lern-dataproducts --vault-password-file /var/lib/jenkins/secrets/vault-pass}
[Pipeline] dir
Running in /var/lib/jenkins/workspace/Deploy/alex/dev/Lern/LernDataProducts/private
[Pipeline] {
[Pipeline] echo
Info: Branch override is enabled
[Pipeline] checkout
The recommended git tool is: NONE
using credential bitbucket-cred
 > git rev-parse --resolve-git-dir /var/lib/jenkins/workspace/Deploy/alex/dev/Lern/LernDataProducts/private/.git # timeout=10
Fetching changes from the remote Git repository
 > git config remote.origin.url bitbucket.org:verlab/veranda-alex-sb-private.git # timeout=10
Using shallow fetch with depth 1
Fetching upstream changes from bitbucket.org:verlab/veranda-alex-sb-private.git
 > git --version # timeout=10
 > git --version # 'git version 2.17.1'
using GIT_SSH to set credentials Veranda Bitbucket Credential (https://bitbucket.org/verlab/veranda-alex-sb-private)
Verifying host key using known hosts file
 > git fetch --no-tags --progress --depth=1 -- bitbucket.org:verlab/veranda-alex-sb-private.git +refs/heads/*:refs/remotes/origin/* # timeout=10
 > git rev-parse origin/release-5.1.0^{commit} # timeout=10
Checking out Revision 182ec55ef65f490f9d7f982b541dfd28ee3ea111 (origin/release-5.1.0)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f 182ec55ef65f490f9d7f982b541dfd28ee3ea111 # timeout=10
Commit message: " we have updated ecr token"
 > git rev-list --no-walk 182ec55ef65f490f9d7f982b541dfd28ee3ea111 # timeout=10
[Pipeline] }
[Pipeline] // dir
[Pipeline] sh
+ rsync -Lkr /var/lib/jenkins/workspace/Deploy/alex/dev/Lern/LernDataProducts/private/ansible/inventory/dev/Lern/common.yml /var/lib/jenkins/workspace/Deploy/alex/dev/Lern/LernDataProducts/private/ansible/inventory/dev/Lern/hosts /var/lib/jenkins/workspace/Deploy/alex/dev/Lern/LernDataProducts/private/ansible/inventory/dev/Lern/secrets.yml /var/lib/jenkins/workspace/Deploy/alex/dev/Lern/LernDataProducts/ansible/inventory/env/
+ [ -f /var/lib/jenkins/workspace/Deploy/alex/dev/Lern/LernDataProducts/ansible/inventory/env/kubernetes.yaml ]
+ ansible-playbook -i /var/lib/jenkins/workspace/Deploy/alex/dev/Lern/LernDataProducts/ansible/inventory/env/hosts --extra-vars remote=spark /var/lib/jenkins/workspace/Deploy/alex/dev/Lern/LernDataProducts/ansible/lpa_lern-data-products_deploy.yml --tags lern-dataproducts --vault-password-file /var/lib/jenkins/secrets/vault-pass
[WARNING]: Invalid characters were found in group names but not replaced, use
-vvvv to see details

PLAY [spark] *******************************************************************

TASK [Gathering Facts] *********************************************************
ok: [172.16.1.34]

TASK [lern-data-products-deploy : Ensure aws blob storage container exists] ****
skipping: [172.16.1.34]

TASK [lern-data-products-deploy : Unarchive Ed Data Products] ******************
ok: [172.16.1.34]

TASK [lern-data-products-deploy : Copy configuration file] *********************
fatal: [172.16.1.34]: FAILED! => {
    "changed": false
}

MSG:

AnsibleUndefinedVariable: 'dict object' has no attribute 'db_password'

PLAY RECAP *********************************************************************
172.16.1.34                : ok=2    changed=0    unreachable=0    failed=1    skipped=1    rescued=0    ignored=0   

[Pipeline] }
[Pipeline] // stage
[Pipeline] }

[Pipeline] // ansiColor
[Pipeline] }
[Pipeline] // stage
[Pipeline] }

[Pipeline] // ansiColor
[Pipeline] ansiColor
[Pipeline] {

[Pipeline] stage
[Pipeline] { (slack_notify)
[Pipeline] sh
+ git rev-parse --short HEAD
[Pipeline] sh
+ git name-rev --name-only HEAD
+ rev
+ cut -d / -f1
+ rev
[Pipeline] sh
+ echo Deploy/alex/dev/Lern/LernDataProducts
[Pipeline] echo
Could not find env specific Slack channel. Check for global slack channel..
[Pipeline] echo
Could not find slack environment variable. Skipping slack notification..
[Pipeline] }
[Pipeline] // stage
[Pipeline] }

[Pipeline] // ansiColor
[Pipeline] ansiColor
[Pipeline] {

[Pipeline] stage
[Pipeline] { (email_notify)
[Pipeline] sh
+ echo Deploy/alex/dev/Lern/LernDataProducts
[Pipeline] echo
Could not find env specific email group. Check for global email group..
[Pipeline] echo
Could not find global email group variable. Skipping email notification..
[Pipeline] }
[Pipeline] // stage
[Pipeline] }

[Pipeline] // ansiColor
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
ERROR: script returned exit code 2
Finished: FAILURE

**Comment by tsprasath** [comment]
@srikanth-6520  Please check ```db_password``` is present in the ansible variables of private repo

**Reply by srikanth-6520** [reply]
@tsprasath "db_password" is present in private repo but we have given an empty value, so please provide "db_password" value.

**Reply by tsprasath** [reply]
Use Postgres password

**Reply by srikanth-6520** [reply]
@tsprasath we have used Postgres password as value , but again it's getting fail with same error.
db_password: "{{ dp_vault_pgdb_admin_password }}"

TASK [lern-data-products-deploy : Copy configuration file] *********************
fatal: [172.16.1.34]: FAILED! => {
    "changed": false
}

MSG:

AnsibleUndefinedVariable: 'dict object' has no attribute 'db_password'

**Reply by heungheung** [reply]
when doing the templ
  https://github.com/srikanth-6520/data-products/blob/release-5.1.0/ansible/roles/lern-data-products-deploy/tasks/main.yml#L63
it is using common.conf.j2
  https://github.com/srikanth-6520/data-products/blob/release-5.1.0/ansible/roles/lern-data-products-deploy/templates/common.conf.j2#L224
and the object postgres should have the db_password attr, which is defined in the group var
  https://github.com/srikanth-6520/data-products/blob/release-5.1.0/ansible/inventory/env/group_vars/all.yml#L54
since it is Lern
  https://github.com/project-sunbird/sunbird-devops/blob/release-5.1.0/private_repo/ansible/inventory/dev/Lern/secrets.yml
it will be soft link to KP
  https://github.com/project-sunbird/sunbird-devops/blob/release-5.1.0/private_repo/ansible/inventory/dev/KnowledgePlatform/secrets.yml#L65

You will need to have your priv repo KP secret dp_vault_pgdb_password var to setup with the analytic db password for the inv.

**Reply by gandham-santhosh** [reply]
@srikanth-6520  please let us know if this is resolved?

**Reply by srikanth-6520** [reply]
@gandham-santhosh we have fixed the issue temporarily by forking the repo and change the cloud commands.

**Reply by gandham-santhosh** [reply]
@srikanth-6520 can you please raise PR for the same?

**Comment by mattkerr7r** [comment]
good
